lib/iris/fileformats/um/_fast_load_structured_fields.py

```diff
@@ -89,15 +87,12 @@ def data(self):
         if not self._structure_calculated:
             self._calculate_structure()
         if self._data_cache is None:
-            data_arrays = [as_lazy_data(f._data, chunks=f._data.shape)
-                           for f in self.fields]
-            vector_dims_list = list(self.vector_dims_shape)
-            vector_dims_list.reverse()
-            self._data_cache = data_arrays
-            for size in vector_dims_list:
-                self._data_cache = [da.stack(self._data_cache[i:i+size]) for i
-                                    in range(0, len(self._data_cache), size)]
-            self._data_cache, = self._data_cache
+            stack = np.empty(np.prod(self.vector_dims_shape), 'object')
```

Comment:
I think it would be neater + clearer to iterate with a multidimensional index here, thus avoiding the ```np.prod(shape)``` and the subsequent ```stack.reshape```.
Your magic friend for that is `np.ndindex(shape)`, which gives you indices to all elements of a multidimensional array, as used [here for example](https://github.com/SciTools/biggus/blob/master/biggus/_init.py#L1812)
I think that gets you something like ...
```
stack = np.empty(self.vector_dims_shape, 'object')
for field, nd_index in zip(fields, np.ndindex(self.vector_dims_shape)):
    stack[nd_index] = as_lazy_data(field._data, chunks=field._data.shape)
self._data_cache = multidim_lazy_stack(stack)
```